### ISSUE TITLE
Display `max_epoch=xxx reached` message when EarlyStopping triggers

### ISSUE THREAD

**Issue by GoldenStain** [Lightning-AI/pytorch-lightning#21031]
### Bug description

When EarlyStopping callback signals Trainer to stop, the message displayed is always wrong. 
## How to reproduce
Any case where EarlyStopping is triggered can reproduce this.
## Cause
In fit_loop.done, the order of if statements are wrong.
```python
    @property
    def done(self) -> bool:
        """Evaluates when to leave the loop."""
        if self.max_batches == 0:
            rank_zero_info("`Trainer.fit` stopped: No training batches.")
            return True

        # TODO: Move track steps inside training loop and move part of these condition inside training loop
        stop_steps = _is_max_limit_reached(self.epoch_loop.global_step, self.max_steps)
        if stop_steps:
            rank_zero_info(f"`Trainer.fit` stopped: `max_steps={self.max_steps!r}` reached.")
            return True

        # `processed` is increased before `on_train_epoch_end`, the hook where checkpoints are typically saved.
        # we use it here because the checkpoint data won't have `completed` increased yet
        assert isinstance(self.max_epochs, int)
        stop_epochs = _is_max_limit_reached(self.epoch_progress.current.processed, self.max_epochs)
        if stop_epochs:
            # in case they are not equal, override so `trainer.current_epoch` has the expected value
            self.epoch_progress.current.completed = self.epoch_progress.current.processed
            rank_zero_info(f"`Trainer.fit` stopped: `max_epochs={self.max_epochs!r}` reached.")
            return True

        if self.trainer.should_stop and self._can_stop_early:
            rank_zero_debug("`Trainer.fit` stopped: `trainer.should_stop` was set.")
            return True

        return False
```
## Solution
Modify the order of if statements.

### What version are you seeing the problem on?

v2.5

### How to reproduce the bug

```python

```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.5.0):
#- PyTorch Version (e.g., 2.5):
#- Python version (e.g., 3.12):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
```

</details>

**Comment by SkafteNicki** [comment]
Hi @GoldenStain, can you provide a bit more details what you mean by the message being always wrong? I quickly tried to put together an example, and it correctly reports that "Monitored metric train_loss did not improve in the last 1 records. Best score: 0.969. Signaling Trainer to stop.` because early stopping was triggered.

```
> python main.py 
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
You are using a CUDA device ('NVIDIA GeForce RTX 4070 Laptop GPU') that has Tensor Cores. To properly utilize them, you should set `torch.set_float32_matmul_precision('medium' | 'high')` which will trade-off precision for performance. For more details, read https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html#torch.set_float32_matmul_precision
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]

  | Name  | Type   | Params | Mode 
-----------------------------------------
0 | layer | Linear | 33     | train
-----------------------------------------
33        Trainable params
0         Non-trainable params
33        Total params
0.000     Total estimated model params size (MB)
1         Modules in train mode
0         Modules in eval mode
/home/nsde/miniforge3/envs/pytorch-lightning/lib/python3.11/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:425: The 'train_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=19` in the `DataLoader` to improve performance.
Epoch 0:   0%|                                                                                                                                                          | 0/4 [00:00<?, ?it/s]/home/nsde/Documents/pytorch-lightning/main.py:31: Using a target size (torch.Size([32])) that is different to the input size (torch.Size([32, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.
Epoch 0:  75%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████▌                                    | 3/4 [00:00<00:00, 18.16it/s]/home/nsde/Documents/pytorch-lightning/main.py:31: Using a target size (torch.Size([4])) that is different to the input size (torch.Size([4, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.
Epoch 0: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 23.80it/s]Metric train_loss improved. New best score: 0.969
Epoch 1: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 274.71it/s]
Monitored metric train_loss did not improve in the last 1 records. Best score: 0.969. Signaling Trainer to stop.
Epoch 1: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 252.57it/s]
```

Expand for testing script:
<details>

```python
import os
import torch
from torch import nn
from torch.utils.data import DataLoader, Dataset
import pytorch_lightning as pl
from pytorch_lightning.callbacks import EarlyStopping

# Dummy dataset
class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index], torch.tensor(0)  # Dummy target

    def __len__(self):
        return self.len

# Simple model
class SimpleModel(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = nn.Linear(32, 1)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        x, y = batch
        loss = nn.functional.mse_loss(self(x), y.float())
        self.log("train_loss", loss + self.current_epoch * 10)  # Simulate increasing loss
        return loss

    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=0.01)

# Main function to test EarlyStopping
def main(): 
    # Set up data
    train_loader = DataLoader(RandomDataset(32, 100), batch_size=32)

    # Model
    model = SimpleModel()

    # EarlyStopping callback
    early_stopping = EarlyStopping(monitor="train_loss", patience=1, mode="min", verbose=True)

    # Trainer
    trainer = pl.Trainer(max_epochs=10, callbacks=[early_stopping])

    # Fit the model
    trainer.fit(model, train_loader)

if __name__ == "__main__":
    main()
``
</details>

**Comment by deependujha** [comment]
Hi @GoldenStain ,

Thanks for reporting this.

> [!IMPORTANT]  
> When opening an issue, please share a **minimal reproducible example** in [Lightning Studio](https://lightning.ai/) so we can run it, confirm the behavior and work on a fix as quickly as possible.

Just to clarify, does this happen only when `max_epochs` and `EarlyStopping` occur at the same time, or even when `max_epochs` hasn’t been reached?

Thanks!

**Comment by GoldenStain** [comment]
> Hi [@GoldenStain](https://github.com/GoldenStain), can you provide a bit more details what you mean by the message being always wrong? I quickly tried to put together an example, and it correctly reports that "Monitored metric train_loss did not improve in the last 1 records. Best score: 0.969. Signaling Trainer to stop.` because early stopping was triggered.
> 
> ```
> > python main.py 
> GPU available: True (cuda), used: True
> TPU available: False, using: 0 TPU cores
> HPU available: False, using: 0 HPUs
> You are using a CUDA device ('NVIDIA GeForce RTX 4070 Laptop GPU') that has Tensor Cores. To properly utilize them, you should set `torch.set_float32_matmul_precision('medium' | 'high')` which will trade-off precision for performance. For more details, read https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html#torch.set_float32_matmul_precision
> LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
> 
>   | Name  | Type   | Params | Mode 
> -----------------------------------------
> 0 | layer | Linear | 33     | train
> -----------------------------------------
> 33        Trainable params
> 0         Non-trainable params
> 33        Total params
> 0.000     Total estimated model params size (MB)
> 1         Modules in train mode
> 0         Modules in eval mode
> /home/nsde/miniforge3/envs/pytorch-lightning/lib/python3.11/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:425: The 'train_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=19` in the `DataLoader` to improve performance.
> Epoch 0:   0%|                                                                                                                                                          | 0/4 [00:00<?, ?it/s]/home/nsde/Documents/pytorch-lightning/main.py:31: Using a target size (torch.Size([32])) that is different to the input size (torch.Size([32, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.
> Epoch 0:  75%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████▌                                    | 3/4 [00:00<00:00, 18.16it/s]/home/nsde/Documents/pytorch-lightning/main.py:31: Using a target size (torch.Size([4])) that is different to the input size (torch.Size([4, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.
> Epoch 0: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 23.80it/s]Metric train_loss improved. New best score: 0.969
> Epoch 1: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 274.71it/s]
> Monitored metric train_loss did not improve in the last 1 records. Best score: 0.969. Signaling Trainer to stop.
> Epoch 1: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 252.57it/s]
> ```
> 
> Expand for testing script:

Soory for the late reply ! Let me provide a example.

**Comment by GoldenStain** [comment]
Now I have a concrete description of how to reproduce the bug on top.
And I already have the solution right now, when can I create a PR?

**Comment by bhimrazy** [comment]
@GoldenStain, Please feel free to open a PR.
> It would also be great if you could share a reproducible example here.

**Comment by GoldenStain** [comment]
> [@GoldenStain](https://github.com/GoldenStain), Please feel free to open a PR.
> 
> > It would also be great if you could share a reproducible example here.

Thank you for the reply! I just wanted to mention that I've already included a reproducible example at the top of this issue, in the main description. Please let me know if you need any additional details or have trouble accessing it.
